File: 1549-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.java
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        PriorityQueue<Integer> min=new PriorityQueue<Integer>();
        PriorityQueue<Integer> max=new PriorityQueue<Integer>(Collections.reverseOrder());
        int n=nums.length;
        int i=0;
        int j=0;
        int len=0;
        while(j<n){
            min.add(nums[j]);
            max.add(nums[j]);
            while(i<j && max.peek()-min.peek()>limit){
                min.remove(nums[i]);
                max.remove(nums[i]);
                i++;
                // System.out.println(max.peek()-min.peek());
            }
            j++;
            len=Math.max(len,j-i);
            // break;
        }
        return len;
    }
}

File: 1549-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cpp
// class Solution {
// public:
//     int longestSubarray(vector<int>& nums, int limit) {
        
//     }
//     #include <queue>
// #include <vector>
// #include <algorithm>

class Solution {
public:
    int longestSubarray(std::vector<int>& nums, int limit) {
        std::multiset<int> minSet;
        std::multiset<int, std::greater<int>> maxSet;
        int n = nums.size();
        int i = 0;
        int j = 0;
        int len = 0;
        
        while (j < n) {
            minSet.insert(nums[j]);
            maxSet.insert(nums[j]);
            
            while (i < j && *maxSet.begin() - *minSet.begin() > limit) {
                minSet.erase(minSet.find(nums[i]));
                maxSet.erase(maxSet.find(nums[i]));
                i++;
            }
            
            j++;
            len = std::max(len, j - i);
        }
        
        return len;
    }
};

// };